File: Java/LoanPattern.java
interface QueryExecutor {
    public abstract execute(Statement stmt);
}

class SomeQueryExecutor {
    public execute(Statement stmt) {
        String query = /* ... */;
        stmt.executeQuery(query);
        /* ... */
    }
}


public void executeSql(String connection, QueryExecutor executor) {
    Connection conn = Connection(connection);
    Statement stmt = conn.createStatement();
    try {
        executor.execute(stmt);
    } finally {
        stmt.close();
        conn.close();
    }
}

File: scala/LoanPattern.scala
def executeSql(connection: String)(block: Statement => Unit) {
  val conn = Connection(connection)
  val stmt = conn.createStatement()
  try {
    block(stmt)
  } finally {
    stmt.close()
    conn.close()
  }
}

val connection: String = /* ... */
executeSql(connection) { statement =>
  val query: String = /* ... */
  statement.executeQuery(query)
  /* ... */
}
